server/src/main/java/org/apache/druid/client/DirectDruidClient.java

```diff
@@ -152,21 +153,23 @@ public Sequence<T> run(final QueryPlus<T> queryPlus, final ResponseContext conte
   {
     final Query<T> query = queryPlus.getQuery();
     QueryToolChest<T, Query<T>> toolChest = warehouse.getToolChest(query);
-    boolean isBySegment = QueryContexts.isBySegment(query);
+    boolean isBySegment = query.context().isBySegment();
     final JavaType queryResultType = isBySegment ? toolChest.getBySegmentResultType() : toolChest.getBaseResultType();
 
     final ListenableFuture<InputStream> future;
     final String url = scheme + "://" + host + "/druid/v2/";
     final String cancelUrl = url + query.getId();
 
     try {
-      log.debug("Querying queryId[%s] url[%s]", query.getId(), url);
+      log.debug("Querying queryId [%s] url [%s]", query.getId(), url);
 
       final long requestStartTimeNs = System.nanoTime();
-      final long timeoutAt = query.getQueryContext().getAsLong(QUERY_FAIL_TIME);
-      final long maxScatterGatherBytes = QueryContexts.getMaxScatterGatherBytes(query);
+      final QueryContext queryContext = query.context();
+      // Will NPE if the value is not set.
```

Comment:
This is a tricky one, hence the comment. The call to `queryContext.getLong(QUERY_FAIL_TIME)` looks up the value, and converts the value to `Long`. If the value is unset, we get a `null` `Long`. When we convert `(long) null` we get a null pointer exception.

In this case, that's OK because `QUERY_FAIL_TIME` is set by Druid itself, so we know it is set correctly. This is the wrong way to get user-provided values: we have to deal with a missing value, or provide a default.

The comment simply means that we thought about the issues, and if a developer sees an NPE, then check what changed in the code because the value should be set.